malfeasance/handler.go

```diff
@@ -92,6 +114,38 @@ func updateMetrics(tp types.Proof) {
 	}
 }
 
+func (h *Handler) validateHareEligibility(ctx context.Context, nodeID types.NodeID, gossip *types.MalfeasanceGossip) error {
+	logger := h.logger.WithContext(ctx).WithFields(nodeID)
+	if gossip == nil || gossip.Eligibility == nil {
+		logger.Fatal("invalid input")
+	}
+
+	emsg := gossip.Eligibility
+	eNodeID := types.BytesToNodeID(emsg.PubKey)
+	if nodeID != eNodeID {
+		logger.With().Warning("mismatch node id", log.Stringer("eligibility", eNodeID))
+		return fmt.Errorf("mismatch node id")
+	}
+	eligible, err := h.validator.Validate(ctx, emsg.Layer, emsg.Round, h.committeeSize, nodeID, emsg.Eligibility.Proof, emsg.Eligibility.Count)
```

Comment:
i just noticed that this is quite expensive and i don't understand why it needs to verify eligibility

it will have to load atxs in the past, do you know if it caches them? if so, when does it drop them from cache?

i thought that it is sufficient for this method to check if two vrf signatures are from the same identity and it can be done without loading lots of data

Reply:
i am quite confident that checking full eligibility is a bug

imagine that hare active set depends on the tortoise. lets say online nodes are using active set X and generate proofs that will pass this validation. later tortoise reaches different consensus (e.g self-healing) and active set is now Y, now nodes that received proof won't be able to verify eligibility. as a consequence identity will be malicious on some nodes, and not malicious on other nodes

Reply:
thank you. hare oracle currently caches the active set for 5 epochs.

the intention here is for live hare rounds. when malicious identity sends a hare message, the rule is for a node to replace the hare message with malfeasance proof, but attach the hare eligibility so the network will still complete the hare instance with the f+1-k rule. 

as such, there should be the same amount of validation before and after this PR. the honest hare eligibilities will be validated in hare/broker.go gossip handler, the dishonest/know equivocator's hare eligibilities will be validated here.

but you are correct that this is a bug, as i didn't check for whether the hare instance for the gossiped eligibility is running.

what do you think that i just unconditionally passed this to the hare, and let hare broker decide whether it should check for full eligibility? @dshulyak

Reply:
i realized that i misunderstood what is going on, but sending msg to hare is indeed looks questionable. i would try to avoid such code

Reply:
ok. thanks. i will pass the hare's gossip handler to the malfeasance handler instead.